bookkeeper-server/src/test/java/org/apache/bookkeeper/client/api/BookKeeperApiTest.java

```diff
@@ -0,0 +1,363 @@
+                }
+            }
+            {
+                Map<String, byte[]> customMetadata = new HashMap<>();
+                customMetadata.put("test", "test".getBytes(StandardCharsets.UTF_8));
+                try (WriteAdvHandle writer
+                        = result(bkc.newCreateLedgerOp()
+                                .withAckQuorumSize(1)
+                                .withWriteQuorumSize(2)
+                                .withEnsembleSize(3)
+                                .withDigestType(DigestType.MAC)
+                                .withCustomMetadata(customMetadata)
+                                .withPassword("password".getBytes(StandardCharsets.UTF_8))
+                                .makeAdv()
+                                .execute());) {
+
+                    long entryId = 0;
+                    byte[] data = "foo".getBytes(StandardCharsets.UTF_8);
+                    writer.write(entryId++, ByteBuffer.wrap(data)).get();
+                    writer.write(entryId++, ByteBuffer.wrap(data)).get();
+                    writer.write(entryId++, Unpooled.wrappedBuffer(data)).get(1, TimeUnit.MINUTES);
+
+                    writer.write(entryId++, ByteBuffer.wrap(data)).get(1, TimeUnit.MINUTES);
+                    long expectedEntryId = writer.write(entryId++, ByteBuffer.wrap(data)).get(1, TimeUnit.MINUTES);
+                    assertEquals(expectedEntryId, writer.getLastAddConfirmed());
+                }
+            }
+        }
+        {
+            Map<String, byte[]> customMetadata = new HashMap<>();
+            customMetadata.put("test", "test".getBytes(StandardCharsets.UTF_8));
+            try (WriteAdvHandle writer
+                    = result(bkc.newCreateLedgerOp()
+                            .withAckQuorumSize(1)
+                            .withWriteQuorumSize(2)
+                            .withEnsembleSize(3)
+                            .withDigestType(DigestType.MAC)
+                            .withCustomMetadata(customMetadata)
+                            .withPassword("password".getBytes(StandardCharsets.UTF_8))
+                            .makeAdv()
+                            .withLedgerId(1234)
+                            .execute());) {
+
+                assertEquals(1234, writer.getId());
+
+                long entryId = 0;
+                byte[] data = "foo".getBytes(StandardCharsets.UTF_8);
+                writer.write(entryId++, ByteBuffer.wrap(data)).get();
+                writer.write(entryId++, ByteBuffer.wrap(data)).get();
+                writer.write(entryId++, Unpooled.wrappedBuffer(data)).get();
+                writer.write(entryId++, ByteBuffer.wrap(data)).get();
+                long expectedEntryId = writer.write(entryId++, ByteBuffer.wrap(data)).get();
+                assertEquals(expectedEntryId, writer.getLastAddConfirmed());
+            }
+        }
+    }
+
+    @Test
+    public void testOpenLedger() throws Exception {
```

Comment:
break this into multiple smaller test cases.

Reply:
done

Reply:
1) can you just break this test into multiple test cases? testOpenLedgerUnauthorized(), testOpenLedgerDigestUnmatched, testOpenLedgerNoSuchLedger, ... It is much clear than putting everything into one single test.

2) you don't need to write data. line 189 to line 192 provide the entries for read.